### ISSUE TITLE
struct constructor doesn't check nested types

### ISSUE THREAD

**Issue by blampe** [uber/thriftrw-python#33]
I don't think we're recursing into lists/maps to reconstruct the native type:

```python
 def test_constructor_behavior_with_nested_types(loads):
     Struct = loads('''struct Struct {
         1: optional list<string> strings;
     }''').Struct

     with pytest.raises(TypeError) as exc_info:
         Struct(strings=[1])
```
```
E   Failed: DID NOT RAISE
```

**Comment by abhinav** [comment]
This will be checked during serialization/deserialization.

**Comment by blampe** [comment]
This makes interacting with the object awkward, though, especially when re-hydrating it from JSON.

**Comment by abhinav** [comment]
Fair. Although we can't drop validation during serialization because that will just fail with cryptic messages when `struct.pack` fails. If we're okay with the added cost, I see no problem with doing it in the constructor too.

**Comment by abhinav** [comment]
Fixed by 0.4.0
